include/CXXGraph/Graph/Algorithm/Pow_impl.hpp

```diff
@@ -0,0 +1,189 @@
+/***********************************************************/
+/***      ______  ____  ______                 _         ***/
+/***     / ___\ \/ /\ \/ / ___|_ __ __ _ _ __ | |__	     ***/
+/***    | |    \  /  \  / |  _| '__/ _` | '_ \| '_ \	 ***/
+/***    | |___ /  \  /  \ |_| | | | (_| | |_) | | | |    ***/
+/***     \____/_/\_\/_/\_\____|_|  \__,_| .__/|_| |_|    ***/
+/***                                    |_|			     ***/
+/***********************************************************/
+/***     Header-Only C++ Library for Graph			     ***/
+/***	 Representation and Algorithms				     ***/
+/***********************************************************/
+/***     Author: ZigRazor ***/
+/***	 E-Mail: zigrazor@gmail.com 				     ***/
+/***********************************************************/
+/***	 Collaboration: ----------- 				     ***/
+/***********************************************************/
+/***	 License: AGPL v3.0 ***/
+/***********************************************************/
+
+#ifndef __CXXGRAPH_POW_IMPL_H__
+#define __CXXGRAPH_POW_IMPL_H__
+
+#pragma once
+
+#include "CXXGraph/Graph/Graph_decl.h"
+
+template <typename T>
+std::vector<std::vector<T>> matMult(std::vector<std::vector<T>> &a, std::vector<std::vector<T>> &b) {
+    static_assert(std::is_same<T, unsigned long long>::value || std::is_same<T, double>::value,
+                  "Type T must be either unsigned long long or double");
+
+    int n = (int)a.size(); // N x N matrix
+    std::vector<std::vector<T>> res(n, std::vector<T>(n, 0));
+
+    // O(n^3) matrix multiplication
+    for (int i = 0; i < n; i++) {
+        for (int j = 0; j < n; j++) {
+            for (int k = 0; k < n; k++) {
+                res[i][j] += a[i][k] * b[k][j];
+            }
+        }
+    }
+
+    return res;
+}
+
+template <typename T>
+std::vector<std::vector<T>> exponentiation(std::vector<std::vector<T>> &mat, unsigned int k) {
+    // typechecking
+    static_assert(std::is_same<T, unsigned long long>::value || std::is_same<T, double>::value,
+                  "Type T must be either unsigned long long or double");
+
+    int n = (int)mat.size();
+    std::vector<std::vector<T>> res(n, std::vector<T>(n, 0)); 
+
+    // build identity matrix    
+    for (int i = 0; i < n; i++) res[i][i] = 1;
+
+    // fast exponentation!
+    while (k) {
+        if (k & 1) res = matMult(res, mat);
+        mat = matMult(mat, mat);
+        k >>= 1;
+    }
+
+    return res;
+}
+
+namespace CXXGraph {
+    
+template <typename T>
+const PowAdjResult Graph<T>::powAdjMatrix(unsigned int k) const {
```

Comment:
Why two specific methods? I think that one method overloaded two times would be better.

Reply:
Okay! I'll just have the user pass an enum value to distinguish. Or, would you prefer they pass the actual matrices themselves to the function? I just want to make sure I'm being consistent with the rest of the repo's style.

Reply:
`virtual const PowAdjResult matrixPow(unsigned int k, AdjacencyMatrix adj) const`
with usage like:
`graph.matrixPow(2, graph.getAdjMatrix());`

This feels wrong to me... should I make a function w/ is not a member of the CXXGraph class? That also feels wrong...

Reply:
Hi, sorry for the delay, I'm very busy these weeks. For me the best thing would be to implement them as free functions, not as Graph methods. But we can hear what @ZigRazor thinks.

Reply:
Sounds good! No worries!

Reply:
Hey @ZigRazor 
Just following up here - would like to get this wrapped up soon! Thoughts on implementing this as a free function?

Reply:
Sorry for the delay.
Yes, the best choice for me is to  implement them as a free functions.
Thank you